### PR TITLE
1.x abstract ipc frontend

### PR DESCRIPTION
**Pull request by flihp**
Should be the finishing touches on the abstractions to encapsulate IPC front-ends.

### CODE REVIEW

File: src/ipc-frontend-dbus.c [hunk 1]
```diff
@@ -0,0 +1,698 @@
...
+ * request from a client to create a new connection with the tabrmd. This
+ * requires a few things be done:
+ * - Create a new ID (uint64) for the connection.
+ * - Create a new Connection object getting the FDs that must be returned
+ *   to the client.
+ * - Build up a dbus response to the client with their connection ID and
+ *   send / receive FDs.
+ * - Send the response message back to the client.
+ * - Insert the new Connection object into the ConnectionManager.
+ * - Notify the CommandSource of the new Connection that it needs to
```
Comment (lugong) [inline src/ipc-frontend-dbus.c]: This action has been removed from the code, so the description should be deleted too.

File: src/ipc-frontend-dbus.c [hunk 2]
```diff
@@ -0,0 +1,698 @@
...
+ * IpcFrontend base class. It causes the IpcFrontendDbus object to connect
+ * to the D-Bus instance provided in the constructor while claiming the name
+ * provided in the same.
+ * This function registers several callbacks with the GDbus machinery. A
+ * reference to the IpcFrontendDbus parameter (self) is passed as data to
+ * these callbacks.
+ */
+void
+ipc_frontend_dbus_connect (IpcFrontendDbus *self,
+                          GMutex         *init_mutex)
```
Comment (lugong) [inline src/ipc-frontend-dbus.c]: adjust formatting

File: src/ipc-frontend-dbus.h [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+#define IPC_FRONTEND_DBUS_CLASS(klass)     (G_TYPE_CHECK_CLASS_CAST    ((klass), TYPE_IPC_FRONTEND_DBUS, IpcFrontendDbusClass))
+#define IS_IPC_FRONTEND_DBUS(obj)          (G_TYPE_CHECK_INSTANCE_TYPE ((obj),   TYPE_IPC_FRONTEND_DBUS))
+#define IS_IPC_FRONTEND_DBUS_CLASS(klass)  (G_TYPE_CHECK_CLASS_TYPE    ((klass), TYPE_IPC_FRONTEND_DBUS))
+#define IPC_FRONTEND_DBUS_GET_CLASS(obj)   (G_TYPE_INSTANCE_GET_CLASS  ((obj),   TYPE_IPC_FRONTEND_DBUS, IpcFrontendDbusClass))
+
+GType           ipc_frontend_dbus_get_type   (void);
+
+IpcFrontendDbus* ipc_frontend_dbus_new        (GBusType           bus_type,
+                                             gchar const       *bus_name,
+                                             ConnectionManager *connection_manager,
```
Comment (lugong) [inline src/ipc-frontend-dbus.h]: adjust formatting.

File: src/ipc-frontend-dbus.h [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+                                             guint              max_trans,
+                                             Random            *random);
+void            ipc_frontend_dbus_connect    (IpcFrontendDbus    *self,
+                                             GMutex            *init_mutex);
```
Comment (lugong) [inline src/ipc-frontend-dbus.h]: adjust formatting.

File: src/ipc-frontend.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+#define TYPE_IPC_FRONTEND             (ipc_frontend_get_type       ())
+#define IPC_FRONTEND(obj)             (G_TYPE_CHECK_INSTANCE_CAST ((obj),   TYPE_IPC_FRONTEND, IpcFrontend))
+#define IS_IPC_FRONTEND(obj)          (G_TYPE_CHECK_INSTANCE_TYPE ((obj),   TYPE_IPC_FRONTEND))
+#define IPC_FRONTEND_CLASS(klass)     (G_TYPE_CHECK_CLASS_CAST    ((klass), TYPE_IPC_FRONTEND, IpcFrontendClass))
+#define IS_IPC_FRONTEND_CLASS(klass)  (G_TYPE_CHECK_CLASS_TYPE    ((klass), TYPE_IPC_FRONTEND))
+#define IPC_FRONTEND_GET_CLASS(obj)   (G_TYPE_INSTANCE_GET_CLASS  ((obj),   TYPE_IPC_FRONTEND, IpcFrontendClass))
+
+GType               ipc_frontend_get_type              (void);
+void                ipc_frontend_connect               (IpcFrontend  *self,
+                                                       GMutex      *mutex);
```
Comment (lugong) [inline src/ipc-frontend.h]: adjust formatting.

File: src/ipc-frontend.h [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+typedef void (*IpcFrontendDisconnect) (IpcFrontend *self);
+
+struct _IpcFrontend {
+    GObject             parent;
+    GMutex             *init_mutex;
+};
+
+struct _IpcFrontendClass {
+    GObjectClass         parent;
+    IpcFrontendConnect    connect;
```
Comment (lugong) [inline src/ipc-frontend.h]: formatting issue.

File: src/ipc-frontend.h [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+#include "connection.h"
+
+G_BEGIN_DECLS
+
+typedef struct _IpcFrontend      IpcFrontend;
+typedef struct _IpcFrontendClass IpcFrontendClass;
+
+typedef void (*IpcFrontendConnect)    (IpcFrontend *self,
+                                      GMutex     *mutex);
```
Comment (lugong) [inline src/ipc-frontend.h]: formatting issue.

File: src/ipc-frontend.c [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+/*
+ * This function is boiler plate. It is used to invoke the function pointed
+ * to by the 'connect' member data / function pointer in the class object.
+ * Derived classes should set this pointer in their class initializer. This
+ * function will use this function pointer to invoke the right function in
+ * derived classes when they're being referenced through this abstract type.
+ */
+void
+ipc_frontend_connect (IpcFrontend *self,
+                     GMutex     *init_mutex)
```
Comment (lugong) [inline src/ipc-frontend.c]: adjust formatting.

File: src/ipc-frontend-dbus.c [hunk 9]
```diff
@@ -0,0 +1,698 @@
...
+    PROP_CONNECTION_MANAGER,
+    PROP_MAX_TRANS,
+    PROP_RANDOM,
+    N_PROPERTIES
+};
+static GParamSpec *obj_properties[N_PROPERTIES] = { NULL };
+
+static void
+ipc_frontend_dbus_set_property (GObject      *object,
+                               guint         property_id,
```
Comment (lugong) [inline src/ipc-frontend-dbus.c]: formatting issue.

File: src/ipc-frontend-dbus.c [hunk 10]
```diff
@@ -0,0 +1,698 @@
...
+        g_object_ref (self->random);
+        break;
+    default:
+        G_OBJECT_WARN_INVALID_PROPERTY_ID (object, property_id, pspec);
+        break;
+    }
+}
+static void
+ipc_frontend_dbus_get_property (GObject    *object,
+                               guint       property_id,
```
Comment (lugong) [inline src/ipc-frontend-dbus.c]: adjust formatting.

File: src/ipc-frontend-dbus.c [hunk 11]
```diff
@@ -0,0 +1,698 @@
...
+                             TYPE_RANDOM,
+                             G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY);
+    g_object_class_install_properties (object_class,
+                                       N_PROPERTIES,
+                                       obj_properties);
+}
+
+IpcFrontendDbus*
+ipc_frontend_dbus_new (GBusType           bus_type,
+                      gchar const       *bus_name,
```
Comment (lugong) [inline src/ipc-frontend-dbus.c]: formatting issue.

File: src/ipc-frontend-dbus.c [hunk 12]
```diff
@@ -0,0 +1,698 @@
...
+ * through the invocation parameter. Mix the two together using xor and
+ * return the result through the id_pid_mix out parameter.
+ * NOTE: if an error occurs then a response is sent through the invocation
+ * to the client and FALSE is returned to the caller.
+ *
+ * Returns FALSE on error, TRUE otherwise.
+ */
+static gboolean
+generate_id_pid_mix_from_invocation (IpcFrontendDbus        *self,
+                                     GDBusMethodInvocation *invocation,
```
Comment (lugong) [inline src/ipc-frontend-dbus.c]: formatting issue.

File: src/ipc-frontend-dbus.c [hunk 13]
```diff
@@ -0,0 +1,698 @@
...
+ *   the ConnectionManager.
+ * - If the connection has a command being processed by the tabrmd then it's
+ *   removed from the processing queue.
+ * - If the connection has a command being processed by the TPM then the
+ *   request to cancel the command will be sent down to the TPM.
+ * - If the connection has no commands outstanding then an error is
+ *   returned.
+ */
+static gboolean
+on_handle_cancel (TctiTabrmd           *skeleton,
```
Comment (lugong) [inline src/ipc-frontend-dbus.c]: formatting issue.

File: src/tabrmd.c [hunk 14]
```diff
@@ -538,6 +94,14 @@ signal_handler (gpointer user_data)
 
     return G_SOURCE_CONTINUE;
 }
+
+static void
+on_ipc_frontend_disconnect (IpcFrontend *ipc_frontend,
+                           GMainLoop  *loop)
```
Comment (lugong) [inline src/tabrmd.c]: formatting issue.

File: src/tabrmd.c [hunk 15]
```diff
@@ -574,32 +139,31 @@ init_thread_func (gpointer user_data)
...
-    if (data->manager == NULL)
+    connection_manager = connection_manager_new(data->options.max_connections);
+    if (connection_manager == NULL)
         g_error ("failed to allocate connection_manager");
-    g_debug ("ConnectionManager: 0x%" PRIxPTR, (uintptr_t)data->manager);
+    g_debug ("ConnectionManager: 0x%" PRIxPTR, (uintptr_t)connection_manager);
+    /* setup IpcFrontend */
+    data->ipc_frontend =
+        IPC_FRONTEND (ipc_frontend_dbus_new (data->options.bus,
+                                            data->options.dbus_name,
```
Comment (lugong) [inline src/tabrmd.c]: adjust formatting.

File: src/tabrmd.c [hunk 16]
```diff
@@ -574,32 +139,31 @@ init_thread_func (gpointer user_data)
...
+                                            data->random));
+    if (data->ipc_frontend == NULL) {
+        g_error ("failed to allocate IpcFrontend object");
+    }
+    g_signal_connect (data->ipc_frontend,
+                      "disconnected",
+                      (GCallback) on_ipc_frontend_disconnect,
+                      data->loop);
+    ipc_frontend_connect (data->ipc_frontend,
+                         &data->init_mutex);
```
Comment (lugong) [inline src/tabrmd.c]: adjust formatting.

File: src/tabrmd.c [hunk 17]
```diff
@@ -827,15 +392,13 @@ main (int argc, char *argv[])
     g_info ("g_main_loop_run done, cleaning up");
     g_thread_join (init_thread);
     /* cleanup glib stuff first so we stop getting events */
-    g_bus_unown_name (gmain_data.dbus_name_owner_id);
-    if (gmain_data.skeleton != NULL)
-        g_object_unref (gmain_data.skeleton);
-    /* tear down the command processing pipeline */
+    ipc_frontend_disconnect (gmain_data.ipc_frontend);
+    g_object_unref (gmain_data.ipc_frontend);
+        /* tear down the command processing pipeline */
```
Comment (lugong) [inline src/tabrmd.c]: adjust formatting.

File: test/ipc-frontend-dbus_unit.c [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+
+#include <setjmp.h>
+#include <cmocka.h>
+
+#include "ipc-frontend-dbus.h"
+
+static int
+ipc_frontend_dbus_setup (void **state)
+{
+    IpcFrontendDbus *ipc_frontend_dbus = NULL;
```
Comment (lugong) [inline test/ipc-frontend-dbus_unit.c]: adjust formatting.

File: test/ipc-frontend-dbus_unit.c [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+    gint ret = 0;
+
+    random = random_new ();
+    ret = random_seed_from_file (random, "/dev/urandom");
+    assert_int_equal (ret, 0);
+
+    connection_manager = connection_manager_new (100);
+
+    ipc_frontend_dbus = ipc_frontend_dbus_new (IPC_FRONTEND_DBUS_TYPE_DEFAULT,
+                                             IPC_FRONTEND_DBUS_NAME_DEFAULT,
```
Comment (lugong) [inline test/ipc-frontend-dbus_unit.c]: formatting issue.

File: test/ipc-frontend_unit.c [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+}
+static void
+test_ipc_frontend_finalize (GObject *obj)
+{
+    if (test_ipc_frontend_parent_class != NULL)
+        G_OBJECT_CLASS (test_ipc_frontend_parent_class)->finalize (obj);
+}
+static void
+test_ipc_frontend_connect (TestIpcFrontend *self,
+                          GMutex         *init_mutex)
```
Comment (lugong) [inline test/ipc-frontend_unit.c]: formatting issue.

File: test/ipc-frontend_unit.c [hunk 21]
```diff
@@ -0,0 +1,237 @@
...
+    assert_true (IS_IPC_FRONTEND (*state));
+    assert_true (IS_TEST_IPC_FRONTEND (*state));
+}
+/*
+ * This callback function is used by the ipc_frontend_event_test to set a
+ * boolean flag when the signal is emitted.
+ */
+static void
+ipc_frontend_on_disconnected (IpcFrontend *ipc_frontend,
+                             bool       *called_flag)
```
Comment (lugong) [inline test/ipc-frontend_unit.c]: formatting issue.

File: test/ipc-frontend_unit.c [hunk 22]
```diff
@@ -0,0 +1,237 @@
...
+}
+/*
+ * This test exercises the 'disconnected' event emitted by the IpcFrontend
+ * abstract class.
+ */
+static void
+ipc_frontend_event_test (void **state)
+{
+    IpcFrontend     *ipc_frontend = NULL;
+    bool            called_flag = false;
```
Comment (lugong) [inline test/ipc-frontend_unit.c]: adjust formatting.

File: src/ipc-frontend-dbus.c (hunk 1 above)
Comment (flihp) [reply src/ipc-frontend-dbus.c]: This whole comment needed to be cleaned up. Thanks for pointing this out.

**Comment by flihp** [comment]
Thanks for getting the review done so quick @lugong & apologies for the sloppy formatting. Should be all fixed up. I'll merge it once the build shows up green.

**Pull request merged**